### DISCUSSION TITLE
Add an icon to the footer

### DISCUSSION THREAD

**Discussion by bluefander** [kdevo/osprey-delight#19] [Q&A] [answered]
Hi

I would like to add a soundcloud icon to the footer:
https://fontawesome.com/v5.15/icons/soundcloud

I thought it should be one of those since it is what fontello-codes.css indicates to me.

In the _icons.scss file I wrote:

```
.icon-soundcloud:before { content: '\f1be'; }
```

And in the config.yaml:

```
- type: soundcloud
  url: myurl
```
But it's not yet displayed in the footer. What is missing?

**Comment by kdevo** [comment, marked as answer]
Hi,

icons are currently only available as a minified/customized icon font in `master` branch. This font needs a rebuild in case of any additional icons than [mentioned in the config](https://github.com/kdevo/osprey-delight/blob/9b1fea2b45dc73e9ee45b477ea3011cb879de510/assets/fontello/config.json#L8), which ensures minimal loading times by only providing a subset of icons. Unfortunately, this also means it's not enough to add the unicode code point to the `_icons.scss` as you did for the SoundCloud icon.

But there's good news: In `dev` branch (think of it as "bleeding edge"), I have already integrated [my solution for dynamically embedding SVG icons](https://github.com/hugo-mods/icons). Hence, you've 3 possibilities to add your SoundCloud icon as far as I can see:

1. Easiest: Check out the `dev` branch, run `hugo mod get -u` and add `soundcloud: fab soundcloud` to `data/icons.yml`. Upside: You'll also profit from the latest other features. Downside: Features might not be as stable/documented than usual.
2. Hard, but correct way for v4: Use [Fontello](https://fontello.com/) by uploading aforementioned config, add SoundCloud, and overwrite the icon font.
3. Impractical, but depends on your time horizon: Wait until the new features get merged into `master`. 

Hope it helps!

BTW: I am kind of curious what your background on using the theme is and I want to learn more about the users of my theme :smile: 
Are you a musician? Do you want to use it as a landing page? Where did you get to know from it? If you want, you can drop me a link to your current WIP version for some feedback.

Have a good day!

**Reply by bluefander** [reply]
Sorry, very much of a noob to github.

I am trying option 1.
So first I should do this, I guess. I made a repo in my own profile called "icons" and ran this in terminal:
```
hugo mod init github.com/bluefander/icons
```

I receive an error:
```
Error: failed to init modules: exec: "go": executable file not found in $PATH
```

Don't know what to do here.


Yes, I am a musician and I am working on a website for a band of mine.
I have bit of a fascination with coding (like trying to learn SuperCollider and Processing) so Hugo resonated with me and I found your theme in https://themes.gohugo.io/.
Once I might feel satisfied with it I might show it. Already abused your theme quite heavily xD

**Reply by kdevo** [reply]
No need for creating an own repo for the icons. 

The following should work:

1. Install [golang](https://golang.org/) for your OS (this is needed for the [Hugo module functionality](https://gohugo.io/hugo-modules/) to work)
2. Initialize your site as a module (example): `hugo mod init github.com/bluefander/my-awesome-site`
3. Add the icon module to your site's config.yaml:
```yaml
//...
module:
  imports:
  - path: github.com/hugo-mods/icons
```
4. Run `hugo mod get -u` to fetch the icon module (this will download the complete FontAwesome icons and might take a while)

Hope I didn't forget anything.

**Reply by kdevo** [reply]
Cool with the band site! Would definitely check it out and can probably give some tips. :smile:

**Reply by bluefander** [reply]
Ok. Step 1-3 seemed to work.

I am not quite sure about step 4. I don't have a file icon.yml in the data folder. There's now a folder called "resources" in the data folder, but it doesn't contain any files.

**Reply by kdevo** [reply]
You can create this file by yourself, it will supplement (**not** replace) [the internal icons.yaml](https://github.com/kdevo/osprey-delight/blob/dev/data/icons.yaml)

**Reply by bluefander** [reply]
Ok, added that.

But now I am completely confused.
If I do:
```
hugo server --themesDir ../..
Error: module "github.com/hugo-mods/icons" not found; either add it as a Hugo Module or store it in "/home/bluefander/hugo/themes".: module does not exist
```

In my config.yaml I wrote: 
```
module:
  imports:
  - path: github.com/hugo-mods/icons
```

**Reply by kdevo** [reply]
This usually happens when you did not init your site as a module.
Where did you execute the ` hugo mod init github.com/bluefander/my-awesome-site`? This needs to happen in your site's directory (where your `config.yaml` is). If it succeeded you'll notice two new files:

- go.mod (contains `module github.com/bluefander/my-awesome-site` in the first line)
- go.sum (contains hashes for file integrity)

If the files are there, could you post the output of the `go.mod`?

**Reply by bluefander** [reply]
Oh sorry, I executed "hugo mod init github.com/bluefander/my-awesome-site" in the wrong directory.
Now the initialisation seemed to work properly. I see go.mod and go.sum in the directory where my config.yaml is located.

In my go.mod I can see:
```
module github.com/bluefander/my-awesome-site

go 1.16

require github.com/hugo-mods/icons v0.0.0-20210529220701-c35284b03de7 // indirect
```

Now "hugo mod get -u" seemed to work properly too. It downloaded some fonts. But where are they located?

I added "soundcloud: fab soundcloud" to my data/icons.yml, but I can't yet see the soundcloud icon on my website.

**Reply by kdevo** [reply]
Is there any warning/output from the icon module when you run `hugo server`?
Are you sure that you're using the dev branch (check with `git branch` in the theme's dir)?

This might not be necessary, but to exclude any possible error, please name the file icons.y**a**ml. Output of `ls` from within your site directory would help, too.

**Reply by bluefander** [reply]
Oh shit... I misunderstood something about the dev branch -.-

But now the dev branch should be working, steps 1-4 completed. I am receiving an error when running "hugo server --themesDir ../.."

```
Start building sites … 
ERROR 2021/06/13 20:35:51 mod-icons: could not find icon 'icon-pulse-fast' by any search method
WARN 2021/06/13 20:35:51 Performance: inlined JS is 6799KB
Built in 144 ms
Error: Error building site: "/home/bluefander/hugo/themes/osprey-delight-dev/exampleSite/content/about.md:11:54": failed to render shortcode "icon": failed to process shortcode: execute of template failed: template: shortcodes/icon.html:2:3: executing "shortcodes/icon.html" at <partial "icon" (dict "context" . "name" (.Get 0) "loader" $loader)>: error calling partial: execute of template failed: template: partials/icon.html:53:19: executing "partials/icon.html" at <partial "icon-resolve" (dict "icon" $icon)>: error calling partial: execute of template failed: template: partials/icon-resolve.html:16:21: executing "partials/icon-resolve.html" at <partial "icon-meta" $iconName>: error calling partial: "/tmp/hugo_cache/modules/filecache/modules/pkg/mod/github.com/hugo-mods/icons@v0.0.0-20210529220701-c35284b03de7/layouts/partials/icon-meta.html:48:13": execute of template failed: template: partials/icon-meta.html:48:13: executing "partials/icon-meta.html" at <index $entry.styles 0>: error calling index: index of untyped nil
```

What am I missing now?


If I comment that out the site builds successfully...
To the very right there is a place where I would expect the soundcloud icon to be and I can find a link when moving my mouse over it, but it isn't displayed.
![footer](https://user-images.githubusercontent.com/85159577/121818483-b1c41100-cc87-11eb-8814-023c704d3990.png)

**Reply by kdevo** [reply]
Remove the `icon-pulse-fast` from your about.md which I suppose is copied from `exampleSite`. Correct?

This icon is not available anymore when using the icon mod, therefore the line...

`ERROR 2021/06/13 20:35:51 mod-icons: could not find icon 'icon-pulse-fast' by any search method`

is outputted.

**Reply by kdevo** [reply]
Could you show me the "Social" section of your config.yaml and content of data/icons.yaml? Is there a yellow `WARN` in console besides the `Performance: inlined JS is 6799KB`?

We'll get this to work :smile:
I've tested it and the icon is displayed on my dummy site.

Please double check that the icon is indeed in data/icons.yaml in your site's directory.
If that does not work, I would suggest that you try to [edit the existing icons.yaml](https://github.com/kdevo/osprey-delight/blob/dev/data/icons.yaml).

**Reply by bluefander** [reply]
config.yaml:

```
  Social:
  - type: twitter # the icon to use. can be anything mentioned in "assets/sass/_icons.scss" 
    url: ""
  - type: linkedin 
    url: ""
  - type: github
    url: github = "https://github.com/kdevo/osprey-delight"
  - type: stackoverflow  
    url: ""
  - type: spotify
    url: https://open.spotify.com/artist/5NhBOdJh6T1tgtM5Au1Opl
  - type: soundcloud
    url: https://soundcloud.com/user-274629117
```

icons.yaml:

```
#some stuff before...
# Social
stackoverflow: fab stack-overflow
linkedin: fab linkedin
github: fab github
medium: fab medium
instagram: fab instagram
twitter: fab twitter
soundcloud: fab soundcloud
```

Only see something like this when starting hugo:

```
Start building sites … 
WARN 2021/06/13 20:45:24 Performance: inlined JS is 6799KB

                   | EN  
```

**Reply by kdevo** [reply]
Looks correct. Did you restart the server after editing icons.yaml?
Could you try changing the `twitter` icon to `fab soundcloud` for debugging purposes?

**Reply by bluefander** [reply]
I restarted it quite a few times.

Tried to change some of the #Social icons in icons.yaml to "fab soundcloud". The effect is that I can "see" a link when moving with the mouse over it but there's no icon displayed

I checked icons.woff with gnome-font-viewer. I can't see a soundcloud icon there if that's relevant.

**Reply by kdevo** [reply]
No, that's not relevant anymore. The file is not even in the dev branch anymore: https://github.com/kdevo/osprey-delight/tree/dev

**Reply by kdevo** [reply]
Not sure what's going wrong there without examining all the files/directory structure.
One other way to "remote debug" this: Right-click on one of the icons and click "Inspect". This will open your browser's dev tools, and what you see there (e.g. provide a screenshot) could help.

**Reply by kdevo** [reply]
When looking at the screenshot in https://github.com/kdevo/osprey-delight/discussions/19#discussioncomment-865562 I think you somehow mixed (or merged) the two branches together, as the old LinkedIn icon has rounded up and the new version has more square-like edges (take a look at my site and compare). On your screenshot, the old version seems to be shown. But I can't tell that for sure, if you want further support, I suggest two things:

1. Take a screenshot of the development tools as described in last post or send me the rendered HTML (right-click and then "view page source").
2. Share your work on GitHub so that I can take a look.

Maybe it also helps if you start over from a clean `dev` state and apply the instructions to the `exampleSite` of the theme.
Otherwise, I think I could add an icon example to the exampleSite tomorrow.

**Reply by bluefander** [reply]
Yeah, I guess that's most likely true that I unintendendly mixed the two -.-
I am trying to start with a clean dev. Can't be that difficult 0o

Thanks a lot for all the help and patience!


I tried with a fresh dev. I am getting these kind of errors:

```
hugo server --themesDir ../..
Start building sites … 
ERROR 2021/06/13 22:01:23 render of "page" failed: execute of template failed: template: _default/single.html:8:7: executing "_default/single.html" at <partial "body-top" .>: error calling partial: "/home/bluefander/desktop/test/osprey-delight/layouts/partials/body-top.html:60:11": execute of template failed: template: partials/body-top.html:60:11: executing "partials/body-top.html" at <partial "icon" "menu">: error calling partial: partial "icon" not found
```

**Reply by kdevo** [reply]
Hi @bluefander 

sure, no problem!
This time, it's because the icon module is missing. If you follow the instructions in https://github.com/kdevo/osprey-delight/discussions/19#discussioncomment-864324 it should work.

Didn't have time yet to adjust the exampleSite yet, but I think you're very close.

**Reply by bluefander** [reply]
Now I really feel like being quite close xD
In my terminal I get:
```
hugo server -D
Error: module "github.com/hugo-mods/icons" not found;
```

Which doesn't make much sense to me since I could see how "hugo mod get -u" downloaded hugo-mods. Just don't get where the stuff is located...
Anyway, if you have the nerves you could find the current state of the theme here:
https://github.com/bluefander/osprey_icons/blob/main/osprey-delight_icons.zip

**Reply by bluefander** [reply]
Hmm... do you have any idea what I am missing?

**Reply by kdevo** [reply]
Hi,

I've added a fully-fledged example for custom icons to the `exampleSite`, please have a look at https://github.com/kdevo/osprey-delight/commit/1e4d8598a95dcb526f299ad3ec635c6b34d6415e ([dev branch](https://github.com/kdevo/osprey-delight/tree/dev/exampleSite)). I've followed [my own instructions](https://github.com/kdevo/osprey-delight/discussions/19#discussioncomment-864324) for this to verify that they work.

PS: I think it would help if you would search for a basic git tutorial that suits you, where you learn some more about the mechanisms, branches etc of git. Git can be a powerful tool for tracking changes, even if you're not a developer, so your time will not be wasted. Uploading a zip file is not the "git way", as it makes it pretty hard for it to track the changes (this is a binary file for git, and its best at doing its work for files it can easily read). Additional benefit: It's also easier for others to look at your changes (like myself). All this is not meant offending of course, just an advice!

**Reply by kdevo** [reply]
> Just don't get where the stuff is located...

I believed this is cached somewhere internally by either go (which I suspect) or Hugo, that's why they are not in the project, as the modules are also meant to be accessible in other projects.

> Which doesn't make much sense to me since I could see how "hugo mod get -u" downloaded hugo-mods.

I think you've forgotten to init (step 2) [as per my instructions](https://github.com/kdevo/osprey-delight/discussions/19#discussioncomment-864324).

**Reply by bluefander** [reply]
I am very sorry... I had some completely stupid misunderstandings about how the directory structure worked in hugo. And that caused some problems. It's logical after all -.-

For sure your 1-4 steps worked out for me, too.

Yeah, I am aware that I have no idea how to use Git properly xD
So I appreciate your advice!
Thanks a lot!

**Reply by kdevo** [reply]
No problem, happy to help :+1:
